### PR TITLE
sampling: remove duplicated code for probability distribution access

### PR DESCRIPTION
**Pull request by mscheong01**
since #5625, function `llama_sampling_probability_distribution` with duplicated code of `llama_sampling_sample` was added to access the probability distribution used for sampling.
This PR is an attempt to remove such code.

### CODE REVIEW

File: common/sampling.cpp [hunk 1]
```diff
@@ -323,8 +269,10 @@ static llama_token_data_array llama_sample_probability_distribution_impl(
     // Get a pointer to the logits
     float * logits = llama_get_logits_ith(ctx_main, idx);
 
-    // Declare original_logits at the beginning of the function scope
-    std::vector<float> original_logits;
+    if (apply_grammar && original_logits != nullptr) {
+        // Only make a copy of the original logits if we are not applying grammar checks, not sure if I actually have to do this.
+        *original_logits = new std::vector<float>(logits, logits + llama_n_vocab(llama_get_model(ctx_main)));
```
Comment (slaren) [inline common/sampling.cpp:274]: When is this freed?

Comment (mscheong01) [reply common/sampling.cpp:277]: I added the line that deletes the allocated vector inside `llama_sampling_sample` in [fddd201](https://github.com/ggerganov/llama.cpp/pull/6240/commits/fddd201942f6b62e42ef3884496f735aa6163d3a)

Comment (mscheong01) [reply common/sampling.cpp]: So the former code where I allocate/deallocated original_logits vector from different scopes seemed like bad practice so I made some adjustments in [27f2e85](https://github.com/ggerganov/llama.cpp/pull/6240/commits/27f2e855203e42b96bdb6adaff3adc051dc90fe0)

File: common/sampling.cpp [hunk 2]
```diff
@@ -323,8 +269,10 @@ static llama_token_data_array llama_sample_probability_distribution_impl(
     // Get a pointer to the logits
     float * logits = llama_get_logits_ith(ctx_main, idx);
 
-    // Declare original_logits at the beginning of the function scope
-    std::vector<float> original_logits;
+    if (apply_grammar && original_logits != NULL) {
+        // Only make a copy of the original logits if we are not applying grammar checks, not sure if I actually have to do this.
+        (*original_logits).insert((*original_logits).end(), logits, logits + llama_n_vocab(llama_get_model(ctx_main)));
```
Comment (cebtenzzre) [inline common/sampling.cpp:274]: This should probably just be:
```suggestion
        *original_logits = {logits, logits + llama_n_vocab(llama_get_model(ctx_main))};
```

File: common/sampling.cpp [hunk 3]
```diff
@@ -302,11 +245,13 @@ static llama_token llama_sampling_sample_impl(
...
 
-static llama_token_data_array llama_sample_probability_distribution_impl(
+static llama_token_data_array llama_sampling_configure_token_candidates_impl(
                   struct llama_sampling_context * ctx_sampling,
                   struct llama_context * ctx_main,
                   struct llama_context * ctx_cfg,
-                  const int idx) {
+                  const int idx,
+                  bool apply_grammar,
+                  std::vector<float>* original_logits) {
```
Comment (cebtenzzre) [inline common/sampling.cpp:254]: Fix this in all three locations:
```suggestion
                  std::vector<float> * original_logits) {
```

File: common/sampling.h [hunk 4]
```diff
@@ -131,12 +131,14 @@ llama_token llama_sampling_sample(
         struct llama_context * ctx_cfg,
         int idx = 0);
 
-// returns the probability that token of given id will be sampled
-llama_token_data_array llama_sampling_probability_distribution(
+// Prepares and adjusts the set of token candidates for sampling based on penalties, biases, and sampling parameters.
+llama_token_data_array llama_sampling_configure_token_candidates(
```
Comment (ggerganov) [inline common/sampling.h:135]: Let's go for something shorter:

```suggestion
llama_token_data_array llama_sampling_prepare(
```

**Review by ggerganov — APPROVED** [review]

File: common/sampling.h (hunk 4 above)
Comment (mscheong01) [reply common/sampling.h]: Addressed in [bb38278](https://github.com/ggerganov/llama.cpp/pull/6240/commits/bb38278e6addca5a1affb00334ea6955242c6465) 👍

**Pull request merged**